out_writeapi_test.go

```diff
@@ -38,3 +44,143 @@ func TestFLBPluginRegister(t *testing.T) {
+}
+
+func (m *MockManagedStream) AppendRows(ctx context.Context, data [][]byte, opts ...managedwriter.AppendOption) (*managedwriter.AppendResult, error) {
+	return m.AppendRowsFunc(ctx, data, opts...)
+}
+
+func (m *MockManagedStream) Close() error {
+	return m.managedstream.Close()
+}
+
+func TestFLBPluginFlushCtx(t *testing.T) {
+	var checks StreamChecks
+	var setID interface{}
+	var id int
+
+	mockMS := &MockManagedStream{
+		AppendRowsFunc: func(ctx context.Context, data [][]byte, opts ...managedwriter.AppendOption) (*managedwriter.AppendResult, error) {
+			checks.appendRows = true
+			return nil, nil
+		},
+	}
+
+	originalFunc := getWriter
+	getWriter = func(client *managedwriter.Client, ctx context.Context, projectID string, opts ...managedwriter.WriterOption) (MWManagedStream, error) {
+		return mockMS, nil
+	}
+	defer func() { getWriter = originalFunc }()
+
+	patch1 := monkey.Patch(output.FLBPluginConfigKey, func(plugin unsafe.Pointer, key string) string {
+		return ""
+	})
+	defer patch1.Unpatch()
+
+	patchDescriptor := monkey.Patch(getDescriptors, func(curr_ctx context.Context, managed_writer_client *managedwriter.Client, project string, dataset string, table string) (protoreflect.MessageDescriptor, *descriptorpb.DescriptorProto) {
+		return nil, nil
+	})
+	defer patchDescriptor.Unpatch()
+
+	patchSetContext := monkey.Patch(output.FLBPluginSetContext, func(plugin unsafe.Pointer, ctx interface{}) {
+		log.Println("Ran set context")
+		setID = ctx
+	})
+	defer patchSetContext.Unpatch()
+
+	plugin := unsafe.Pointer(nil)
+	// ctx := unsafe.Pointer(nil)
+	initRes := FLBPluginInit(plugin)
+
+	orgFunc := getContext
+	getContext = func(ctx unsafe.Pointer) int {
+		checks.calledGetContext = true
+		uintptrValueBack := uintptr(ctx)
+		id = int(uintptrValueBack)
+		return id
+	}
+	defer func() { getContext = orgFunc }()
+
+	patchCheckResponses := monkey.Patch(checkResponses, func(curr_ctx context.Context, currQueuePointer *[]*managedwriter.AppendResult, waitForResponse bool) int {
+		checks.calledcheckResponses = true
+		return 0
+	})
+	defer patchCheckResponses.Unpatch()
+
+	patchDecoder := monkey.Patch(output.NewDecoder, func(data unsafe.Pointer, length int) *output.FLBDecoder {
+		checks.createDecoder = true
+		return nil
+	})
+	defer patchDecoder.Unpatch()
+
+	var loopCount int = 0
+	patchRecord := monkey.Patch(output.GetRecord, func(dec *output.FLBDecoder) (ret int, ts interface{}, rec map[interface{}]interface{}) {
+		checks.gotRecord = true
+		if loopCount%2 == 0 {
+			loopCount++
+			return 0, nil, nil
+		} else {
+			loopCount++
+			return 1, nil, nil
+		}
+	})
+	defer patchRecord.Unpatch()
+
+	patchParse := monkey.Patch(parseMap, func(mapInterface map[interface{}]interface{}) map[string]interface{} {
+		checks.calledparseMap = true
+		return nil
+	})
+	defer patchParse.Unpatch()
+
+	patchJTB := monkey.Patch(json_to_binary, func(message_descriptor protoreflect.MessageDescriptor, jsonRow map[string]interface{}) ([]byte, error) {
```

Comment:
Internal function should not be tested.

Reply:
By this do you mean that I should just remove the check that json_to_binary was called or that I should remove the monkey patching completely? The monkey patching is necessary because we still need to mock it so we can "convert" data without actually having any data (so it can actually access AppendRows).

Reply:
Yes I mean remove the monkey patching completely for this internal function.

Could you provide valid JSON data when you call Flush()? Then the conversion to protobuf would also happen. Note that a valid table schema would also have to be provided under Init().

Reply:
Okay! Yeah I'm going to manually have to create data but I can definitely do that. That will also make testing checkResponses easier!

Reply:
Done!